builtin/game/anticheat.lua

```diff
@@ -0,0 +1,288 @@
+						.. " completed digging " .. core.pos_to_string(node_pos)
+						.. "; which is not diggable with tool. not digging.")
+				core.run_anticheat_callbacks(self.player, "dug_unbreakable", {node_pos = node_pos, groups = groups})
+				return true
+			end
+		end
+
+		return false
+	end
+})
+
+core.register_anticheat_object("dug_too_fast", {
+	check = function(self, node_pos, nocheat_time, grab_lag_pool)
+		if not self.enabled then
+			return false
+		end
+
+		local groups = core.registered_items[core.get_node(node_pos).name].groups
+		local tool_capabilities = self.player:get_wielded_item():get_tool_capabilities()
+		local dtime = core.get_dig_params(groups, tool_capabilities).time
+
+		if dtime > 2.0 and nocheat_time * 1.2 > dtime then
+			-- All is good, but grab time from pool; don't care if
+			-- it's actually available
+			grab_lag_pool(dtime, self.player:get_player_name(), "dig")
+			return false
+		elseif grab_lag_pool(dtime, self.player:get_player_name(), "dig") then
+			return false
+		else
+			core.log("info", "Server: NoCheat: " .. self.player:get_player_name()
+							.. " completed digging " .. core.pos_to_string(node_pos)
+							.. "too fast; not digging.")
+			core.run_anticheat_callbacks(self.player, "dug_too_fast", {node_pos = node_pos,
+				dtime = dtime, nocheat_time = nocheat_time})
+			return true
+		end
+	end
+})
+
+core.register_anticheat_object("moved_too_fast", {
+	check = function(self, pos, max_lag_estimate, time_from_last_teleport, last_good_position, grab_lag_pool)
+		if not self.enabled or self.player:get_attach() then
+			return false
+		end
+
+		--[[
+			Check player movements
+
+			NOTE: Actually the server should handle player physics like the
+			client does and compare player's position to what is calculated
+			on our side. This is required when eg. players fly due to an
+			explosion. Altough a node-based alternative might be possible
+			too, and much more lightweight.
+		]]
+
+		local player_max_speed = self:get_max_speed()
+		-- Tolerance. The lag pool does this a bit.
+		--player_max_speed = player_max_speed * 2.5
+
+		local diff = vector.subtract(pos, last_good_position)
+		local d_vert = diff.y
+		diff.y = 0
+		local d_horiz = vector.length(diff)
+		local required_time = d_horiz / player_max_speed
+
+		if d_vert > 0 and d_vert / player_max_speed > required_time then
+			required_time = d_vert / player_max_speed -- Moving upwards
+		end
+
+		if grab_lag_pool(required_time, self.player:get_player_name(), "move") then
+			return false
+		else
+			local LAG_POOL_MIN = 5.0
+			local lag_pool_max = math.max(max_lag_estimate * 2.0, LAG_POOL_MIN)
+			if time_from_last_teleport > lag_pool_max then
+				minetest.log("action", "Player " .. self.player:get_player_name()
+						.. " moved too fast; resetting position")
+				core.run_anticheat_callbacks(self.player, "moved_too_fast", {player_pos = pos,
+					player_max_speed = player_max_speed})
+				return true
+			end
```

Comment:
lots of times in this file:
doesn't need to be in an `else` block since the other block returns from function